### ISSUE TITLE
Is moving mean and variance in batchnorm layer updated properly?

### ISSUE THREAD

**Issue by hiwonjoon** [cbfinn/maml#16]
Hello,

I am trying to reimplement MAML by myself, but I figured it out that implementing batch norm is tricky. Since statistics of data can change over internal iteration (inside loop) by SGD, it is hard to decide how beta and gamma (not used for this case because of ReLU nonlinearity) should be updated. Furthermore, it is rather unclear which means and variances should be tracked for validation. Should it be tracked overall mean and variance across tasks and internal iterations? Or, should the statistics before it is tuned to tasks be measured?

I tried some of the possible strategies, but the result did not match up with the performance reported. The best performance was observed without using batch norm, but it still 95% for Omniglot Dataset for 5-way, 1-shot problem. 

With this consideration, I looked up the code and I found out that moving means and variances are not saved and restored. I could not find:
```  
update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
```
so, I checked the saved model with the code:
```
        # from main.py
        if FLAGS.test_iter > 0:
            model_file = model_file[:model_file.index('model')] + 'model' + str(FLAGS.test_iter)
        if model_file:
            ind1 = model_file.index('model')
            resume_itr = int(model_file[ind1+5:])
            print("Restoring model weights from " + model_file)
            with tf.variable_scope("", reuse=True):
                b = tf.get_variable("model/0/beta",)
                mm = tf.get_variable("model/0/moving_mean",)
                mv = tf.get_variable("model/0/moving_variance")

                print( sess.run([b,mm,mv]) )
```
and the result was
```
Restoring model weights from logs/omniglot5way//cls_5.mbs_32.ubs_1.numstep1.updatelr0.4batchnorm/model2000
[array([-0.01856588, -0.07139841,  0.00564138,  0.01990231, -0.00762643,
       -0.05611767,  0.07031356, -0.03621985,  0.05920702,  0.12788662,
        0.04555263, -0.06217157, -0.07977205,  0.01632672, -0.03578645,
        0.10676555, -0.04455299, -0.0573478 ,  0.52247691, -0.05695038,
        0.14302482, -0.07892933, -0.02123305,  0.01870824,  0.01471483,
       -0.06067625,  0.097821  , -0.05786318,  0.03801388, -0.04843186,
       -0.01786073,  0.0293963 ,  0.56441385,  0.07509601,  0.11491237,
        0.01052142,  0.23142786,  0.03433308, -0.05783347, -0.0444839 ,
        0.02227049, -0.02804896, -0.04594825,  0.05347209, -0.0399643 ,
        0.02923759,  0.1299762 , -0.02817831, -0.0735756 , -0.0284342 ,
       -0.04498725, -0.05203079, -0.04267518, -0.03341504, -0.05648317,
       -0.02747083, -0.03525382,  0.34740165, -0.00822794,  0.03952603,
        0.03410957,  0.29954502, -0.01362322, -0.04790628], dtype=float32), array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.], dtype=float32), array([ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,
        1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,
        1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,
        1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,
        1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.], dtype=float32)]
```

It feels like the only beta variables are updated, so it seems what batchnorm layers do is shifting only.

Any thoughts?

Thanks!

**Comment by hiwonjoon** [comment]
Duplicate of #9,

Sorry for opening new a new ticket. I will close it.

**Comment by cbfinn** [comment]
Ok. And, yes, only beta is being used because the scaling can be handled by the scale of the weights in the following layer.

**Comment by hiwonjoon** [comment]
Thanks for the such an instant response!

You mentioned at #9 that mean and variance of batch can be used instead of the tracked mean and variance, but, does Instance Normalization fit well in this situation? (Well, calculating statistics for each "batch" in the few-shot learning case does make sense in that "batch" is the whole dataset in this case...)

Any thoughts on this?

**Comment by cbfinn** [comment]
You could certainly try instance norm. It would be pretty easy to add, and the statistics only rely on the current example. The implementation would look very similar to the layer norm implementation in the code.
